API for URL Shortening

DevanshCodes opened the pull request:
API for URL Shortening using the Date.now() to make unique short ID and valid URL function to check the validity of URLs.  Storing in DB: longurl, shorturl, urlcode.

DevanshCodes commented:
> @DevanshCodes Looks good!
> You can add functions and server, DB interaction in the controllers.
> Nice work.

Okay, will add. Can you also list what all things are to be added. Adding of function which will take description, search url in it, shorten it using API and then sending the shorten URL, in controllers right? Is there anything else?

vaibhavdaren commented on app/routes/url.js:
```diff
@@ -0,0 +1,43 @@
+function validURL(myURL) {
+    var pattern = new RegExp('^(https?:\\/\\/)?'+ 
+    '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.?)+[a-z]{2,}|'+ 
+    '((\\d{1,3}\\.){3}\\d{1,3}))'+ 
```
store the regex in a string and use that string as regex expression

DevanshCodes replied:
Okay.

vaibhavdaren commented on app/routes/url.js:
```diff
@@ -0,0 +1,43 @@
+router.post('/shorten',async (req,res) => {
+    const longurl =  req.body;
+    var baseurl = req.get('host');
+    const urlcode = Date.now();
```
var/const

DevanshCodes replied:
Okay, will fix that. :)

DevanshCodes commented:
@vaibhavdaren All the changes are done. :)

devesh-verma commented on app/routes/index.js:
```diff
@@ -1,9 +1,25 @@
+    res.json('Server error!')
+  }
+});
+
```
We can have this in a separate file. `urlShortner.js`

DevanshCodes replied:
Done, I've added this in the url.js file in which there was POST Request only.

DevanshCodes replied:
But @devesh-verma this GET request should remain in the Index.js file, if I try to move it in any other file then instead of shortURL being `127.0.0.1:3000/1584954626498` it will become `127.0.0.1:3000/shortUrl/1584954626498` as the handler will change. Or we can add a redirect from the Index file. 
![image](https://user-images.githubusercontent.com/48180687/77305459-08a89800-6d1c-11ea-935f-3c2eab668c69.png) 

Please let me know using which approach I should move forward?

devesh-verma replied:
this is fine, nothing wrong here.

devesh-verma commented on app/models/ShortUrl.js:
```diff
@@ -0,0 +1,17 @@
+const mongoose = require('mongoose')
+
+const ShortUrlSchema = new mongoose.Schema({
```
name this as `urlShortnerSchema`

devesh-verma commented on app/models/ShortUrl.js:
```diff
@@ -0,0 +1,17 @@
+})
+
+const shortURL = mongoose.model('shortURL',ShortUrlSchema);
+module.exports = shortURL;
```
change the file name from `ShortUrl.js` to `UrlShortner.js`

devesh-verma commented on app.js:
```diff
@@ -26,6 +27,7 @@ app.use('/', indexRouter)
 app.use('/auth', authRouter)
 app.use('/user', usersRouter)
 app.use('/post', postRouter)
+app.use('/api', urlRouter)
```
`app.use('/api', urlRouter)` -> `app.use('/shortUrl', shortUrlRouter)`

devesh-verma commented on app.js:
```diff
@@ -9,6 +9,7 @@ const indexRouter = require('./app/routes/index')
 const authRouter = require('./app/routes/auth')
 const usersRouter = require('./app/routes/user')
 const postRouter = require('./app/routes/post')
+const urlRouter = require('./app/routes/url')
```
`const urlRouter = require('./app/routes/url')`  -> `const shortUrlRouter = require('./app/routes/url')`

DevanshCodes commented:
I've pushed the changes with redirecting to /shortUrl.

DevanshCodes commented:
Working Gif
![WorkingAPI](https://user-images.githubusercontent.com/48180687/77306720-14955980-6d1e-11ea-86e5-4a8d28dc6b24.gif)

devesh-verma commented:
@DevanshCodes gif seems to wrong.

devesh-verma commented on app/routes/urlShortner.js:
```diff
@@ -0,0 +1,60 @@
+      }
+    } catch (error) {
+      res.json('Server error!')
+    }
```
this logic needs to go inside controller

devesh-verma commented on app/routes/urlShortner.js:
```diff
@@ -0,0 +1,60 @@
+        }
+    }else{
+        res.json('invalid long url');
+    }
```
this logic too.

devesh-verma commented:
@DevanshCodes also please check why is travis failing for your PR.
take latest pull, as this was fixed by @Rupeshiya

DevanshCodes commented:
> @DevanshCodes gif seems to be wrong.

Wrong in the sense? Should there be more things in the gif?

DevanshCodes commented:
I've moved all the logic to the controllers, also added the comments, and the structure similar to other routes and controllers for uniformity. I've attached another GIF though I'm not clear what was wrong in the already attached gif. Also this was before pulling the latest changes from the repo.
![WorkingExample](https://user-images.githubusercontent.com/48180687/77345432-9c00be00-6d5a-11ea-955f-46bfe03a7aff.gif)

After pulling all the changes from the repo, this error comes when running the app
![ErrorShowing](https://user-images.githubusercontent.com/48180687/77345861-3a8d1f00-6d5b-11ea-9dc3-40fab6daaba1.gif)

Also event after pulling the latest changes, Travis build is still failing. :(

devesh-verma commented:
@DevanshCodes take lates pull now I have fixed this, and do a `npm install
`

DevanshCodes commented:
Done @devesh-verma its working fine now on port 5000.
